modules/gateway/peers.go

```diff
@@ -437,6 +531,23 @@ func (g *Gateway) managedConnect(addr modules.NetAddress) error {
 		conn.Close()
 		return err
 	}
+
+	// Ensure peer is not connection to itself,
+	// the peers are on the same network, and a connection is wanted.
+	if build.VersionCmp(remoteVersion, sessionHandshakeUpgradeVersion) >= 0 {
+		wantConn := true
+		wantConn, err = connectSessionHandshake(conn, g.sessionHeader(wantConn))
+		if err != nil {
+			conn.Close()
+			return err
+		}
+
+		if !wantConn {
+			conn.Close()
+			return errPeerRejectedConn
+		}
+	}
```

Comment:
this should be an explicit `if-else` chain:
```go
if build.VersionCmp(remoteVersion, sessionHandshakeUpgradeVersion) >= 0 {
    // new peer
} else if build.VersionCmp(remoteVersion, handshakeUpgradeVersion) >= 0 {
    // old peer
} else {
    // really old peer
}
```
this requires duplicating some code but it's less prone to bugs.